src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -387,5 +387,81 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
         {
             return p.Designation.Kind() == SyntaxKind.SingleVariableDesignation && p.IsOutDeclaration();
         }
+
+        /// <summary>
+        /// Visits all the ArrayRankSpecifiers of a typeSyntax, invoking an action on each one in turn.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="action"></param>
+        /// <param name="argument">The argument that is passed to the action whenever it is invoked</param>
+        internal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<ArrayRankSpecifierSyntax, TArg> action, in TArg argument)
+        {
+recurse:
+            switch (type.Kind())
+            {
+                case SyntaxKind.ArrayType:
+                    var arrayTypeSyntax = (ArrayTypeSyntax)type;
+                    arrayTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    foreach (var rankSpecifier in arrayTypeSyntax.RankSpecifiers)
+                    {
+                        action(rankSpecifier, argument);
+                    }
+                    break;
+                case SyntaxKind.NullableType:
+                    var nullableTypeSyntax = (NullableTypeSyntax)type;
+                    type = nullableTypeSyntax.ElementType;
+                    goto recurse;
+                case SyntaxKind.PointerType:
+                    var pointerTypeSyntax = (PointerTypeSyntax)type;
+                    type = pointerTypeSyntax.ElementType;
+                    goto recurse;
+                case SyntaxKind.TupleType:
+                    var tupleTypeSyntax = (TupleTypeSyntax)type;
+                    var elementsCount = tupleTypeSyntax.Elements.Count;
+                    if (elementsCount == 0)
+                        break;
+
+                    for (int index = 0; index < elementsCount - 1; index++)
+                    {
+                        var element = tupleTypeSyntax.Elements[index];
+                        element.Type.VisitRankSpecifiers(action, argument);
+                    }
+
+                    type = tupleTypeSyntax.Elements[elementsCount - 1].Type;
+                    goto recurse;
+                case SyntaxKind.RefType:
+                    var refTypeSyntax = (RefTypeSyntax)type;
+                    type = refTypeSyntax.Type;
+                    goto recurse;
+                case SyntaxKind.GenericName:
+                    var genericNameSyntax = (GenericNameSyntax)type;
+                    var argsCount = genericNameSyntax.TypeArgumentList.Arguments.Count;
+                    Debug.Assert(argsCount != 0);
```

Comment:
>Debug.Assert(argsCount != 0); [](start = 20, length = 29)

Is it not possible to violate this assumption by building the node manually?

Reply:
Consider replacing the assert with:
```
                    if (argsCount == 0)
                        break;
```

similar to the way TupleType is handled

---
In reply to: [266938083](https://github.com/dotnet/roslyn/pull/32843#discussion_r266938083) [](ancestors = 266938083)

Reply:
Done